### PR TITLE
Split repo trusted setting

### PR DESCRIPTION
**Pull request by qwerty287**
related to https://github.com/woodpecker-ci/woodpecker/discussions/3758

### CODE REVIEW

File: server/store/datastore/migration/035_split_trusted.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+		if err := dropTableColumns(sess, "repos", "trusted"); err != nil {
+			return err
+		}
+
+		if err := sess.Commit(); err != nil {
+			return err
+		}
+
+		if err := renameColumn(sess, "repos", "trusted_conf", "trusted"); err != nil {
```
Comment (zc-devs) [inline server/store/datastore/migration/035_split_trusted.go:69]: `trusted` column has been deleted at 61 line. Maybe this is not necessary.

**Review by zc-devs — COMMENTED** [review]
> get it into 3.0

**Comment by anbraten** [comment]
I am not really a fan of this split. Although I kinda see the general idea behind it, it complicates the CI (not really interested to move towards Jenkins and similar tools) and it shifts the security guidelines from the agent admin to the repo admin. Therefore multiple agents aren't able to provide different settings.

**Review by 6543 — APPROVED** [review]

**Review by 6543 — CHANGES_REQUESTED** [review]
migration fail related !!!

https://ci.woodpecker-ci.org/repos/3780/pipeline/20192/36

**Review by anbraten — CHANGES_REQUESTED** [review]
Not confined by the benefit of splitting the setting.

**Comment by anbraten** [comment]
> Repo admins can't change this setting if they're not a server admin and usually the server admin is also the agent admin I think.

Isn't that the reason why this is currently an agent config (at least for the kubernetes options). It would only make sense for smaller deployments where server admins are agent admins as well I guess. For larger teams / orgs with or when supporting #267 those options would be quite unimportant.

**Comment by zc-devs** [comment]
> this is currently an agent config (at least for the kubernetes options)
Where?

> this issue is not different from how it works now. Enabling trusted is currently also on repo-level and this PR does not give any way to get more privileges than before

**Comment by zc-devs** [comment]
> You mean the "feature gates"?

I was asking. @anbraten, what did you mean by `agent config`? Could you provide an example?

Agent's feature gates != settings per a repo.

> having it on a repo level additionally can still be helpful as you might have repos that shouldn't be able to use some features

--- 

> it complicates the CI

The same way as `trusted` does.

> not really interested to move towards Jenkins and similar tools

Could you elaborate?

> it shifts the security guidelines from the agent admin to the repo admin

To the server/instance admin, if I understand correctly. The same as `trusted`. And doesn't shift, but adds per repo settings.

**Comment by zc-devs** [comment]
Maybe using [ALTER TABLE DROP COLUMN](https://www.sqlite.org/lang_altertable.html) will help.
`server/store/datastore/migration/common.go:dropTableColumns(...)`

**Comment by zc-devs** [comment]
[Using_official_DROP_COLUMN_syntax.patch](https://github.com/user-attachments/files/16922333/Using_official_DROP_COLUMN_syntax.patch) works for me on your branch
```
root@4296e32370b1:/src/woodpecker# make test-server-datastore-coverage
go test -race -cover -coverprofile datastore-coverage.out -timeout 300s -tags 'test ' go.woodpecker-ci.org/woodpecker/v2/server/store/...
        go.woodpecker-ci.org/woodpecker/v2/server/store         coverage: 0.0% of statements
?       go.woodpecker-ci.org/woodpecker/v2/server/store/types   [no test files]
        go.woodpecker-ci.org/woodpecker/v2/server/store/mocks           coverage: 0.0% of statements
ok      go.woodpecker-ci.org/woodpecker/v2/server/store/datastore       2.236s  coverage: 72.3% of statements
ok      go.woodpecker-ci.org/woodpecker/v2/server/store/datastore/migration     48.395s coverage: 58.8% of statements
root@4296e32370b1:/src/woodpecker#
```

**Comment by 6543** [comment]
> Maybe using [ALTER TABLE DROP COLUMN](https://www.sqlite.org/lang_altertable.html) will help. `server/store/datastore/migration/common.go:dropTableColumns(...)`
> 
> Offtop: I thought, ORM/migration tool should do this magic. Seems, I was wrong :)

ORM dont just drop data ... you have to specify that

**Comment by 6543** [comment]
As for the current structure I'm quite happy as it works (some strings atached...) but to cover the concerns ... an agent admin should be able to optionally overwrite security related things set by the server, if veasable like e.g. deny to mount volumes pointing to non docker volumes e.g. to the host ... 

But that has to be a backemd option as each backend has different security considderations

File: server/store/datastore/migration/022_add_orgs.go [hunk 2]
```diff
@@ -66,15 +66,10 @@ var addOrgs = xormigrate.Migration{
 			}
 		}
 
-		if err := sess.Sync(new(model.Org), new(syncRepo022), new(userV031)); err != nil {
+		if err := sess.Sync(new(model.Org), new(syncRepo022), new(userV031), new(oldSecret022)); err != nil {
```
Comment (6543) [inline server/store/datastore/migration/022_add_orgs.go:69]: Unrelated?

File: server/store/datastore/migration/036_split_trusted.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		if err := sess.Sync(new(repoV035)); err != nil {
+			return fmt.Errorf("sync new models failed: %w", err)
+		}
+
+		var repos []*repoV035
+		if err := sess.Find(&repos); err != nil {
+			return err
+		}
+
+		for _, r := range repos {
```
Comment (6543) [inline server/store/datastore/migration/036_split_trusted.go:48]: You dont have to itterate in codd, let it do the database ...


Where(trusted).Col(x).Update(newModel(datafornew col))

File: server/store/datastore/migration/common.go [hunk 4]
```diff
@@ -173,7 +121,7 @@ func dropTableColumns(sess *xorm.Session, tableName string, columnNames ...strin
 			cols += "DROP COLUMN `" + col + "`"
 		}
 		if _, err := sess.Exec(fmt.Sprintf("ALTER TABLE `%s` %s", tableName, cols)); err != nil {
-			return fmt.Errorf("drop table `%s` columns %v: %w", tableName, columnNames, err)
+			return fmt.Errorf("table `%s`, drop columns %v: %w", tableName, columnNames, err)
```
Comment (6543) [inline server/store/datastore/migration/common.go:124]: Unrelated?

File: pipeline/frontend/yaml/linter/linter_test.go [hunk 5]
```diff
@@ -94,7 +94,12 @@ steps:
 			conf, err := yaml.ParseString(testd.Data)
 			assert.NoError(t, err)
 
-			assert.NoError(t, linter.New(linter.WithTrusted(true)).Lint([]*linter.WorkflowConfig{{
+			assert.NoError(t, linter.New(linter.WithTrusted(linter.TrustedConfiguration{
+				Network: true,
+				Volumes: true,
+
```
Comment (6543) [inline pipeline/frontend/yaml/linter/linter_test.go:100]:
```suggestion
```
?

File: cli/exec/flags.go [hunk 6]
```diff
@@ -184,10 +184,21 @@ var flags = []cli.Flag{
...
+	},
+	&cli.BoolFlag{
+		Sources: cli.EnvVars("CI_REPO_TRUSTED_VOLUMES", "CI_REPO_TRUSTED"),
+		Name:    "repo-trusted-volumes",
+		Usage:   "Set the metadata environment variable \"CI_REPO_TRUSTED_VOLUMES\".",
+	},
+	&cli.BoolFlag{
+		Sources: cli.EnvVars("CI_REPO_TRUSTED_SECURITY", "CI_REPO_TRUSTED"),
+		Name:    "repo-trusted-security",
+		Usage:   "Set the metadata environment variable \"CI_REPO_TRUSTED_SECURITY\".",
```
Comment (6543) [inline cli/exec/flags.go:201]:
```suggestion
		Sources: cli.EnvVars("CI_REPO_TRUSTED_NETWORK"),
		Name:    "repo-trusted-network",
		Usage:   "Set the metadata environment variable \"CI_REPO_TRUSTED_NETWORK\".",
	},
	&cli.BoolFlag{
		Sources: cli.EnvVars("CI_REPO_TRUSTED_VOLUMES"),
		Name:    "repo-trusted-volumes",
		Usage:   "Set the metadata environment variable \"CI_REPO_TRUSTED_VOLUMES\".",
	},
	&cli.BoolFlag{
		Sources: cli.EnvVars("CI_REPO_TRUSTED_SECURITY"),
		Name:    "repo-trusted-security",
		Usage:   "Set the metadata environment variable \"CI_REPO_TRUSTED_SECURITY\".",
```

File: cli/exec/flags.go [hunk 7]
```diff
@@ -184,10 +184,21 @@ var flags = []cli.Flag{
 		Name:    "repo-private",
 		Usage:   "Set the metadata environment variable \"CI_REPO_PRIVATE\".",
 	},
+	// TODO remove CI_REPO_TRUSTED in 4.x
```
Comment (6543) [inline cli/exec/flags.go:187]:
```suggestion
```

**Review by 6543 — APPROVED** [review]

**Comment by anbraten** [comment]
> @anbraten maybe we can just merge this now as it's ready and discuss the general approach again?

Still not really convinced how this would really help users and think it mainly ignores UX while complicating existing options. Especially as it doesn't consider org level agents. Anyways will trust you making this helpful at some point.

File: server/store/datastore/migration/017_split_trusted.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+var splitTrusted = xormigrate.Migration{
+	ID: "split-trusted",
+	MigrateSession: func(sess *xorm.Session) error {
+		if err := sess.Sync(new(repoV035)); err != nil {
+			return fmt.Errorf("sync new models failed: %w", err)
+		}
+
+		var repos []*repoV035
```
Comment (6543) [inline server/store/datastore/migration/017_split_trusted.go:43]: pagination is missing ...

**Review by 6543 — CHANGES_REQUESTED** [review]
migration ...

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@qwerty287 lint err

**Comment by igniscyan** [comment]
> > @anbraten maybe we can just merge this now as it's ready and discuss the general approach again?
> 
> Still not really convinced how this would really help users and think it mainly ignores UX while complicating existing options. Especially as it doesn't consider org level agents. Anyways will trust you making this helpful at some point.

Am I understanding correctly that the VOLUME, NETWORK, and SECURITY flags were agent settings before rather than repo settings?

**Pull request merged**